cni/pkg/plugin/iptables.go

```diff
@@ -55,11 +61,25 @@ func (ipt *iptables) Program(podName, netns string, rdrct *Redirect) error {
 	viper.Set(constants.RedirectDNS, rdrct.dnsRedirect)
 	viper.Set(constants.CaptureAllDNS, rdrct.dnsRedirect)
 	viper.Set(constants.DropInvalid, rdrct.invalidDrop)
-	iptablesCmd := cmd.GetCommand()
-	log.Infof("============= Start iptables configuration for %v =============", podName)
-	defer log.Infof("============= End iptables configuration for %v =============", podName)
-	if err := iptablesCmd.Execute(); err != nil {
+
+	netNs, err := getNs(netns)
+	if err != nil {
+		err = fmt.Errorf("failed to open netns %q: %s", netns, err)
 		return err
 	}
+	defer netNs.Close()
+
+	if err = netNs.Do(func(_ ns.NetNS) error {
```

Comment:
Isn't this unsafe? https://www.weave.works/blog/linux-namespaces-and-go-don-t-mix? It seems like _maybe_ its safer now, but still need to do some more effort to lock things: https://github.com/vishvananda/netns#note

Reply:
Interesting article! The locking mentioned in there is thankfully already handled by the package in the `Do` method. They also provide a few remarks surrounding safety in their [README](https://github.com/containernetworking/plugins/tree/master/pkg/ns). The Istio CNI process should exit fairly quickly, minimizing the risk even further.

This namespace switching can also be found in other popular CNI projects, e.g. [cilium](https://github.com/cilium/cilium/blob/master/plugins/cilium-cni/chaining/generic-veth/generic-veth.go#L57-L64), so it's not completely out there.

Unfortunately, I can't think of a better way to make the code behave the same within a pod and on the host atm. We could provide the IP from the CNI `prevResult` as a parameter to the istio-iptables code, but that would a) be a bigger change and b) also possibly lead to different behavior between the two "modes".

Reply:
Ah interesting. I guess my info is from 2017 :stuck_out_tongue: .

So one concern is we already have logic to run stuff in the network namespace. so now we enter the netns, then from within the netns we enter it again? This probably is technical fine but likely means we have over complicated things -- if its safe to do for this use case we can simplify the rest of the code as well?

How the rest works is we shell out - https://github.com/istio/istio/blob/abaae0893cc14c44edf290e48b9b748d2fcd45d1/tools/istio-iptables/pkg/capture/run.go#L287 - for everything that is done in the pod netns.

So my thinking is either we should do the same thing in both places. I don't think I am educated enough right now to say which one we should do - ciliums reference makes this approach seem plausible.
cc @bianpengyuan @lambdai

Reply:
I think we'd have to nsenter when shell out?  Not sure how do we keep the namespace for a subprocess with this mechanism?

Reply:
I looked at the shell out code and believe it should work fine using this mechanism as well, without `nsenter`. `Cmd.Run` [documents](https://github.com/golang/go/blob/d588f487703e773ba4a2f0a04f2d4141610bff6b/src/os/exec/exec.go#L333-L336) that in a locked thread the network namespace will be inherited for the executed command.

That being said, if you spawn a new goroutine from within `Do` it will not inherit the new namespace and behave unexpectedly. I thought about it some more and would propose the following changes to this pull request:

- namespace switching as implemented here is moved from the plugin to istio-iptables, which would allow us to just wrap the relevant `getLocalIP` call [here](https://github.com/istio/istio/blob/abaae0893cc14c44edf290e48b9b748d2fcd45d1/tools/istio-iptables/pkg/cmd/root.go#L152) if the `CNIMode` flag is set
- if you wish, the other calls can also be wrapped within go by adapting the namespaced [execute](https://github.com/istio/istio/blob/abaae0893cc14c44edf290e48b9b748d2fcd45d1/tools/istio-iptables/pkg/dependencies/implementation.go#L74-L78) instead of using nsenter - maybe in a follow up PR to keep the changeset small on this one?

What do you think @howardjohn & @bianpengyuan?